accepted/2021/install-location-per-architecture.md

```diff
@@ -138,34 +141,56 @@ any specific knowledge of where to find the install location, it would just
 invoke the right apphost.
 
 On all platforms the apphost will have a default location hardcoded
-and so if the installer puts the layout into that location it doesn't have to
+and so if the installer puts the product into that location it doesn't have to
 register that installation. But it would be cleaner if it did so anyway.
-Currently we always register the installation on Windows, but we don't seem
-to do so on Linux and macOS.
+Currently we always register the installation on Windows, but we don't do so
+on Linux and macOS.
 
-*Note: If installers write the file, they should also read it and validate
-that they don't overwrite it with different information. It's unclear yet
-if this is necessary and in which situations.*
+#### Windows installer behavior
 
-#### Support for downlevel apphost
+Pre-.NET 6 Windows installer writes registry keys for all architectures
+(even though it only installs one of them). Ideally the installer would
+only write the key for the architecture it's actually installing.
 
-In order to support running downlevel apphost apps on multi-arch non-Windows
-architectures (currently only applies to macOS), the installer of .NET 6 and above
-should write the configuration file in a specific shape, for example (on macOS arm64):
+#### Linux/macOS installer behavior
 
-```console
-/usr/local/share/dotnet/x64
-arm64=/usr/local/share/dotnet
-x64=/usr/local/share/dotnet/x64
-```
+Pre-.NET 6 Linux/macOS installer doesn't write the `install_location` file.
 
-With the above proposed changes this would guarantee that all apphosts
-would work if the right frameworks are installed.
+Starting with .NET 6, installers should write architecture specific file
+with the product location, in the format `install_location_<arch>`. So for example
+`install_location_x64` (all lower case). The content of the file should be
+a single line with the path to the product.
 
-* 3.1 and 5 apphost would only read the first line, which points to x64 install.
-Since 3.1 and 5 are only supported on x64 on macOS, that is the right location.
-* arm64 6 apphost would react to the `arm64` location
-* x64 6 apphost would react to the `x64` location
+*Note: If installers can write the file, they should also read it and validate
+that they don't overwrite it with different information.*
```

Comment:
Can you clarify this?  Are you saying that whenever any of the 3 files exist, we should leave them alone and not write them at all?  "Verify they don't overwrite with different information" sounds unusual to me: if it isn't different writing is a noop.   I'd like to understand if you're expecting anything more than just skip when it exists. (like testing the path for presence of dotnet.exe) 

We should really decide if we want all/some of these files to be _user_ state or _product_ state.  I understand that the legacy path was _user_ state before, but we're turning it into both _user_ and _product_ state since it suits our purposes to allow old apps to work.  We can avoid stomping on a user value by skipping to write the file at all if it exists.

We should decide if we want the same behavior for the new files.

The downside of having files in ambiguous _user/product_ state is that the installer cannot repair them easily (without encoding more rules about what is valid and thus complicating the install) since it can't distinguish a value written by the user from a default written by the product.

Reply:
This comment was driven by the Windows behavior. On Windows the rule should be (I actually don't know the installer well enough to tell this) that if I already installed for example 3.1 x64 into a custom path (`D:\dotnet`), then if I try to install 5.0 x64, it should install into the same location `D:\dotnet` and not "override" the location in the registry.

Currently we don't do this on Linux/macOS - and maybe we don't want to - not sure.

I honestly always considered both the registry on Windows and this file on Linux/macOS to be a "product state". There could be some users setting this, but it should be VERY rare (unless you're on one of the special distros which wants to install into different location, but those distros probably don't use our packages by default anyway).

Reply:
I tried to make it clearer.

Reply:
> This comment was driven by the Windows behavior. 

Yes, Windows does this, however it's not well advertised.  We definitely don't tell folks to write those keys, they change the path through arguments passed to the bundle.  It coincidentally uses the same key as input in the SDK bundle (though not the runtime).  In these cases it's preserving the path of the single machine-wide installer-managed runtime.

cc @sfoslund On mac we don't support generic retargeting and the mechanism we're using today to retarget to `x64` doesn't lend itself to that since it must have a [static absolute path](https://developer.apple.com/library/archive/documentation/DeveloperTools/Reference/DistributionDefinitionRef/Chapters/Distribution_XML_Ref.html#//apple_ref/doc/uid/TP40005370-CH100-SW47).  We tried to use [relocatable bundles](https://developer.apple.com/library/archive/documentation/DeveloperTools/Reference/DistributionDefinitionRef/Chapters/Distribution_XML_Ref.html#//apple_ref/doc/uid/TP40005370-CH100-SW87) and searches but couldn't get macos to honor them: our best guess is that this is a deprecated feature because it doesn't work as documented and we couldn't find anyone successfully using it.  If we could get those to work, then the installer could implement a javascript search that read the file and returned the value, or a default (with x64 appended as appropriate).  Since we cannot, it doesn't seem plausible to "preserve" custom directory installations on mac.

> but it should be VERY rare 

I think we just need to make the call what we should do in that rare case.  We can't honor it as the single installer managed location as we do on windows.  Do we allow it to remain untouched and let folks have some custom location that's side-by-side?  Or do we overwrite it to point to the installer managed location.

Reply:
Thanks for the explanation. Thinking about it some more - I would probably prefer we overwrite actually. Reasoning: I just explicitly installed for example .NET 6 x64. Immediately after I try to run 6.0 x64 app, but it doesn't work because I had my own config and the installer effectively silently didn't update it. That feels like a bad experience. We just position the config files as "maintained by the installer" and so if somebody modifies them, a new installation overwriting makes sense.

@richlander for another opinion.

Reply:
I changed the text to explicitly state that we overwrite and why.